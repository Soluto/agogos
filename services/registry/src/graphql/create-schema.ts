import gql from "graphql-tag";
import { mergeDocuments } from "./schema-utils";
import { DocumentNode } from "graphql";

type GqlSource = {
  source: string;
  gql: string;
};

export type GqlSources = { [source: string]: string };

function parseGqlSource(gqlSource: GqlSource) {
  try {
    return gql(gqlSource.gql);
  } catch (e) {
    throw {
      message: `Failed to load schema for source -  ${gqlSource.source} - ${
        e.message
      }`,
      schema: gqlSource,
      error: e
    };
  }
}

function mergeAllDocuments(docs: DocumentNode[]) {
  try {
    return mergeDocuments(docs);
  } catch (e) {
    throw {
      message: `Failed to merge schemas - ${e.message}`,
      schemas: docs,
      error: e
    };
  }
}

export const makeGqlDocumentFromGqlSources = (gqlSources: GqlSources) => {
  const documentNodes = Object.entries(gqlSources).map(([source, gql]) =>
    parseGqlSource({ source, gql })
  );
  return mergeAllDocuments(documentNodes);
};
