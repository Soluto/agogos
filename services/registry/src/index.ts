import nconf = require("nconf");
import logger, { loggingMiddleware } from "./logger";
nconf.env("__");

import * as express from "express";
import gqlSchemaRoute from "./routes/schema";
import validateRoute from "./routes/validate";
import { startGrpcServer } from "./sync/sync-grpc-server";

const PORT = process.env.PORT || 4000;

const app = express();

app.use(loggingMiddleware);

app.get("/health", (_: express.Request, res: express.Response) =>
    res.send(true),
);
app.get("/metrics", (_: express.Request, res: express.Response) =>
    res.send(true),
);

app.use("/schema", gqlSchemaRoute);
app.use("/validate", validateRoute);

app.listen({ port: PORT }, () =>
    logger.info(`ðŸš€ HTTP Server ready at http://localhost:${PORT}`),
);

startGrpcServer();
