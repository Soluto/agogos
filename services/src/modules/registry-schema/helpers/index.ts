import proceedPolicy from './policy-helper';

export { proceedPolicy };
