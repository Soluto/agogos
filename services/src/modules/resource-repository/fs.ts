import {ResourceGroup, ResourceRepository} from '.';
import * as envVar from 'env-var';
import {promises as fs} from 'fs';
import * as path from 'path';
import pLimit from 'p-limit';
import * as config from '../config';
import logger from '../logger';
import {FetchLatestResult} from './types';

export class FileSystemResourceRepository implements ResourceRepository {
    protected current?: {mtime: number; rg: ResourceGroup};
    protected policyAttachmentsDirInitialized = false;
    protected policyAttachments: {[filename: string]: Buffer} = {};
    protected policyAttachmentsRefreshedAt?: Date;

    constructor(protected pathToFile: string, protected policyAttachmentsFolderPath: string) {}

    async fetchLatest(): Promise<FetchLatestResult> {
        const stats = await fs.stat(this.pathToFile);

        if (this.current && stats.mtimeMs === this.current.mtime) {
            return {isNew: false, resourceGroup: this.current.rg};
        }

        const contents = await fs.readFile(this.pathToFile, 'utf8');
        const rg = JSON.parse(contents) as ResourceGroup;
        this.current = {mtime: stats.mtimeMs, rg};

        return {isNew: true, resourceGroup: rg};
    }

    async update(rg: ResourceGroup): Promise<void> {
        await fs.writeFile(this.pathToFile, JSON.stringify(rg));
    }

    async writePolicyAttachment(filename: string, content: Buffer): Promise<void> {
        await this.initializePolicyAttachmentsDir();

        const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
        await fs.writeFile(filePath, content);
    }

    public getPolicyAttachment(filename: string): Buffer {
        return this.policyAttachments[filename];
    }

    public async initializePolicyAttachments() {
        try {
            await this.refreshPolicyAttachments();
        } catch (err) {
            logger.fatal({err}, 'Failed fetching fs policy attachments on startup');
            throw err;
        }

        setInterval(async () => {
            try {
                await this.refreshPolicyAttachments();
            } catch (err) {
                logger.error(
                    {err},
                    `Failed refreshing fs policy attachments, last successful refresh was at ${this.policyAttachmentsRefreshedAt}`
                );
            }
        }, config.resourceUpdateInterval);
    }

    private async refreshPolicyAttachments() {
        const newRefreshedAt = new Date();

        const allAttachments = await this.getPolicyAttachmentsList();
        const attachmentsToRefresh = allAttachments
            .filter(a => this.shouldRefreshPolicyAttachment(a))
            .map(a => a.filename);

        if (attachmentsToRefresh.length > 0) {
            const newAttachments = await this.getPolicyAttachments(attachmentsToRefresh);
            newAttachments.forEach(a => (this.policyAttachments[a.filename] = a.content));
        }

        this.policyAttachmentsRefreshedAt = newRefreshedAt;
    }

    private shouldRefreshPolicyAttachment({filename, updatedAt}: {filename: string; updatedAt: Date}) {
        if (!this.policyAttachments[filename]) return true;
        if (!this.policyAttachmentsRefreshedAt) return true;

        return updatedAt > this.policyAttachmentsRefreshedAt;
    }

    private async getPolicyAttachmentsList(): Promise<{filename: string; updatedAt: Date}[]> {
        const filenames = await fs.readdir(this.policyAttachmentsFolderPath);
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const stats = await fs.stat(filePath);

                    const updatedAt = stats.mtime;
                    return {filename, updatedAt};
                });
            })
        );
    }

    private async getPolicyAttachments(filenames: string[]): Promise<{filename: string; content: Buffer}[]> {
        const limit = pLimit(10);

        return Promise.all(
            filenames.map(filename => {
                return limit(async () => {
                    const filePath = path.resolve(this.policyAttachmentsFolderPath, filename);
                    const content = await fs.readFile(filePath);

                    return {filename, content};
                });
            })
        );
    }

    private async initializePolicyAttachmentsDir() {
        if (this.policyAttachmentsDirInitialized) return;

        await fs.mkdir(this.policyAttachmentsFolderPath, {recursive: true});
        this.policyAttachmentsDirInitialized = true;
    }

    static fromEnvironment() {
        const pathToFile = envVar
            .get('FS_RESOURCE_REPOSITORY_PATH')
            .required()
            .asString();
        const policyAttachmentsFolderPath = envVar
            .get('FS_REPOSITORY_POLICY_ATTACHMENTS_FOLDER_PATH')
            .required()
            .asString();

        return new FileSystemResourceRepository(pathToFile, policyAttachmentsFolderPath);
    }
}
