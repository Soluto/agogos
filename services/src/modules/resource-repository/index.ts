export * from './types';

export { ResourceRepository } from './resource-repository';
export { CompositeResourceRepository } from './composite';
export { S3ResourceRepository } from './s3-repository';
export { FileSystemResourceRepository } from './fs-repository';
export { pollForUpdates } from './stream';
export { applyResourceGroupUpdates, applyResourceGroupDeletions } from './actions';
export { default as getResourceRepository } from './get-resource-repository';
