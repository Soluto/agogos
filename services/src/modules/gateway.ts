import * as fastify from 'fastify';
import { ApolloServer, Config } from 'apollo-server-fastify';
import { Observable } from 'rxjs';
import { createGraphQLService } from './graphql-service';
import { RESTDirectiveDataSource } from './directives/rest';
import { ResourceGroup } from './resource-repository';
import { ExportTrackingExtension } from './exports';
import {
  createBasicPolicyPlugin,
  createLoggingPlugin,
  createMetricsPlugin,
  createArgumentInjectionPlugin,
} from './plugins';

export interface GatewayConfig extends Config {
  resourceGroups: Observable<ResourceGroup>;
  fastifyInstance: Pick<fastify.FastifyInstance, 'metrics'>;
}

export function createStitchGateway(config: GatewayConfig) {
  const { resourceGroups, fastifyInstance, ...apolloConfig } = config;
  const gateway = createGraphQLService({ resourceGroups });
  const server = new ApolloServer({
    ...apolloConfig,
    gateway,
    extensions: [() => new ExportTrackingExtension()],
    plugins: [
      createBasicPolicyPlugin,
      createLoggingPlugin(),
      createMetricsPlugin(fastifyInstance),
      createArgumentInjectionPlugin(),
    ],
    subscriptions: false,
    context(request: fastify.FastifyRequest) {
      const ctx = {
        request,
      };
      return ctx;
    },
    dataSources() {
      return {
        rest: new RESTDirectiveDataSource(),
      };
    },
  });

  return {
    server,
    dispose: async () => {
      await server.stop();
      gateway.dispose();
    },
  };
}
