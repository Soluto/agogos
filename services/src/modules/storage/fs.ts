import * as path from 'path';
import { promises as fs } from 'fs';
import pLimit from 'p-limit';
import { Storage, FileStats, listFilesItem } from '.';

export class FileSystemStorage implements Storage {
  async fileStats(filePath: string): Promise<FileStats> {
    const stats = await fs.stat(filePath);

    return { lastModified: stats.mtime };
  }

  async readFile(filePath: string): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { [k in any]: never }): Promise<{ content: Buffer }>;
  async readFile(filePath: string, options: { asString?: true }): Promise<{ content: string }>;
  async readFile(filePath: string, options: { asString?: boolean } = {}): Promise<{ content: Buffer | string }> {
    if (options.asString) return { content: await fs.readFile(filePath, 'utf8') };

    return { content: await fs.readFile(filePath) };
  }

  async writeFile(filePath: string, content: string | Buffer): Promise<void> {
    await fs.writeFile(filePath, content);
  }

  async deleteFile(filePath: string): Promise<void> {
    await fs.unlink(filePath);
  }

  async listFiles(folderPath: string): Promise<listFilesItem[]> {
    const filenames = await fs.readdir(folderPath);
    const limit = pLimit(10);

    return Promise.all(
      filenames.map(filename => {
        return limit(async () => {
          const filePath = path.resolve(folderPath, filename);
          const { lastModified } = await this.fileStats(filePath);

          return { filePath, lastModified };
        });
      })
    );
  }

  async mkdir(folderPath: string, options?: { recursive: true }): Promise<void> {
    await fs.mkdir(folderPath, options);
  }
}
