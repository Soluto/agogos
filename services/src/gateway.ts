import * as fastify from 'fastify';
import * as fastifyMetrics from 'fastify-metrics';

import * as config from './modules/config';
import logger from './modules/logger';
import {handleSignals, handleUncaughtErrors} from './modules/shutdownHandler';
import {createStitchGateway} from './modules/gateway';
import {pollForUpdates, S3ResourceRepository} from './modules/resource-repository';

async function run() {
    logger.info('Stitch gateway booting up...');

    const resourceRepository = new S3ResourceRepository();

    const {server, dispose} = createStitchGateway({
        resourceGroups: pollForUpdates(resourceRepository, config.resourceUpdateInterval),
        tracing: config.enableGraphQLTracing,
        playground: config.enableGraphQLPlayground,
        introspection: config.enableGraphQLIntrospection,
    });

    const app = fastify();
    app.register(fastifyMetrics, {endpoint: '/metrics'});
    app.register(server.createHandler({path: '/graphql'}));
    const address = await app.listen(config.httpPort, '0.0.0.0');
    logger.info({address}, 'Stitch gateway started');

    handleSignals(dispose);
    handleUncaughtErrors();
}

// Only run when file is being executed, not when being imported
if (require.main === module) {
    run();
}
