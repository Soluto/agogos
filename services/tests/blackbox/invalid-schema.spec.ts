import * as fs from 'fs/promises';
import { print } from 'graphql';
import { FastifyInstance } from 'fastify';
import { gql, GraphQLRequest } from 'apollo-server-core';
import { createServer as createGateway } from '../../src/gateway';
import { ResourceGroup, Schema } from '../../src/modules/resource-repository';
import { sleep } from '../helpers/utility';

async function makeDefaultQuery(app: FastifyInstance) {
  const payload: GraphQLRequest = {
    query: print(gql`
      query {
        default
      }
    `),
  };

  const response = await app.inject({
    method: 'POST',
    url: '/graphql',
    payload,
  });
  return response;
}

describe('Invalid schema', () => {
  let app: FastifyInstance;

  beforeAll(async () => {
    const resources: ResourceGroup = {
      schemas: [],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
    };

    await fs.writeFile(process.env.FS_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));

    app = await createGateway();
  });

  afterAll(async () => {
    await app.close();
    await fs.unlink(process.env.FS_RESOURCE_REPOSITORY_PATH!);
  });

  test('Make default query', async () => {
    const response = await makeDefaultQuery(app);

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });

  test('Save invalid schema', async () => {
    const schema: Schema = {
      metadata: {
        namespace: 'blackbox',
        name: 'invalid-schema',
      },
      schema: print(gql`
        type Query {
          foo: WrongType!
        }
      `),
    };

    const resources: ResourceGroup = {
      schemas: [schema],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
    };
    await fs.writeFile(process.env.FS_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));

    const response = await app.inject({
      method: 'POST',
      url: '/updateSchema',
    });
    expect(response.statusCode).toEqual(500);
  });

  test('Make default query again', async () => {
    const response = await makeDefaultQuery(app);

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });

  test('Wait for refresh and make default query', async () => {
    await sleep(500);
    const response = await makeDefaultQuery(app);

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });

  test('Save valid schema', async () => {
    const schema: Schema = {
      metadata: {
        namespace: 'blackbox',
        name: 'invalid-schema',
      },
      schema: print(gql`
        type Query {
          foo: String! @localResolver(value: "FOO")
        }
      `),
    };

    const resources: ResourceGroup = {
      schemas: [schema],
      upstreams: [],
      upstreamClientCredentials: [],
      policies: [],
    };
    await fs.writeFile(process.env.FS_RESOURCE_REPOSITORY_PATH!, JSON.stringify(resources));

    const response = await app.inject({
      method: 'POST',
      url: '/updateSchema',
    });
    expect(response.statusCode).toEqual(200);
  });

  test('Query new field', async () => {
    const payload: GraphQLRequest = {
      query: print(gql`
        query {
          foo
        }
      `),
    };

    const response = await app.inject({
      method: 'POST',
      url: '/graphql',
      payload,
    });

    expect(response.statusCode).toEqual(200);
    expect(response.json().data).toMatchSnapshot();
  });
});
