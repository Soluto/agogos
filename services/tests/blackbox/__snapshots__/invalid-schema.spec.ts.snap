// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Invalid schema Make default query 1`] = `
Object {
  "default": "default",
}
`;

exports[`Invalid schema Make default query again 1`] = `
Object {
  "default": "default",
}
`;

exports[`Invalid schema Query new field 1`] = `
Object {
  "foo": "FOO",
}
`;

exports[`Invalid schema Wait for refresh and make default query 1`] = `
Object {
  "default": "default",
}
`;
